libmamba/include/mamba/util/graph.hpp

```diff
@@ -443,127 +456,241 @@ namespace mamba::util
     template <typename UnaryFunc>
     UnaryFunc DiGraphBase<N, G>::for_each_leaf_id_from(node_id source, UnaryFunc func) const
     {
-        struct LeafVisitor : default_visitor<derived_t>
-        {
-            UnaryFunc& m_func;
-
-            LeafVisitor(UnaryFunc& func)
-                : m_func{ func }
-            {
-            }
-
-            void start_node(node_id n, const derived_t& g)
+        dfs_preorder_nodes_for_each_id(
```

Comment:
I'm not sure about this, but shouldn't the `Leaf ` be replaced with postorder instead? (or maybe the trick was using reverse in the `Root`?).
I think that root/leaf keywords give more insight about this, maybe add a comment if we plan to remove them?

Reply:
> I'm not sure about this, but shouldn't the Leaf be replaced with postorder instead?

It makes no difference, as they are no more nodes to visit after this one and the algorithm backtracks.

> or maybe the trick was using reverse in the Root?

Exactly

>  I think that root/leaf keywords give more insight about this, maybe add a comment if we plan to remove them?

I'm not sure what you mean.

Reply:
> > I think that root/leaf keywords give more insight about this, maybe add a comment if we plan to remove them?
> 
> I'm not sure what you mean.

`LeafVisitor` and `RootVisitor` don't exist anymore, but the naming was maybe helping to understand the purpose of the code? So I suggested adding a comment for that (to explain what it does).

Reply:
> LeafVisitor and RootVisitor don't exist anymore,

Ha yes, the visitor pattern was more verbose so I switch to the preorder function.

> So I suggested adding a comment for that (to explain what it does).

Got it.